modules/basegl/src/processors/imageprocessing/layershader.cpp

```diff
@@ -0,0 +1,280 @@
+ *
+ * Copyright (c) 2023 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/basegl/processors/imageprocessing/layershader.h>
+
+#include <inviwo/core/datastructures/image/layer.h>
+#include <inviwo/core/util/foreacharg.h>
+#include <modules/opengl/shader/shaderresource.h>
+#include <modules/opengl/shader/shaderutils.h>
+#include <modules/opengl/texture/textureunit.h>
+#include <modules/opengl/texture/textureutils.h>
+#include <modules/opengl/image/layergl.h>
+
+namespace inviwo {
+
+namespace detail {
+
+constexpr std::string_view vertexShader = R"(out Fragment {
+    smooth vec4 color;
+    smooth vec2 texCoord;
+} out_vert;
+
+void main() {
+    out_vert.color = in_Color;
+    out_vert.texCoord = in_TexCoord.xy;
+    gl_Position = in_Vertex;
+}
+)";
+
+constexpr std::string_view defaultFrag = R"(#include "utils/structs.glsl"
+#include "utils/sampler2d.glsl"
+
+uniform sampler2D layer;
+uniform ImageParameters layerParameters;
+
+struct FormatScaling {
+    float formatScaling;
+    float formatOffset;
+    float signedFormatScaling;
+    float signedFormatOffset;
+};
+
+uniform FormatScaling destRange = FormatScaling(1.0, 0.0, 1.0, 0.0);
+
+in Fragment {
+    smooth vec4 color;
+    smooth vec2 texCoord;
+} in_frag;
+
+void main() {
+    // access normalized input value, then renormalize with respect to output data range
+    vec4 value = getNormalizedTexel(layer, layerParameters, in_frag.texCoord);
+    FragData0 = value / (1.0 - destRange.formatScaling) - destRange.formatOffset;
+
+    // regular texture access, no output mapping
```

Comment:
remove? make option?

Reply:
This was on purpose to showcase both versions: a normalized one and one with actual data values.